akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala

```diff
@@ -4,215 +4,45 @@
      * Behavior’s type hierarchy. Signals are not transformed.
      *
-     * see also [[Actor.Widened]]
+     * Example:
+     * {{{
+     * Immutable[String] { (ctx, msg) => println(msg); Same }.widen[Number] {
+     *   case b: BigDecimal => s"BigDecimal(&dollar;b)"
+     *   case i: BigInteger => s"BigInteger(&dollar;i)"
+     *   // drop all other kinds of Number
+     * }
+     * }}}
      */
-    def widen[U](matcher: PartialFunction[U, T]): Behavior[U] = Widened(behavior, matcher)
-  }
-
-  private val _nullFun = (_: Any) ⇒ null
-  private def nullFun[T] = _nullFun.asInstanceOf[Any ⇒ T]
-  private implicit class ContextAs[T](val ctx: AC[T]) extends AnyVal {
-    def as[U] = ctx.asInstanceOf[AC[U]]
-  }
-
-  /**
-   * Widen the wrapped Behavior by placing a funnel in front of it: the supplied
-   * PartialFunction decides which message to pull in (those that it is defined
-   * at) and may transform the incoming message to place them into the wrapped
-   * Behavior’s type hierarchy. Signals are not transformed.
-   *
-   * Example:
-   * {{{
-   * Immutable[String] { (ctx, msg) => println(msg); Same }.widen[Number] {
-   *   case b: BigDecimal => s"BigDecimal(&dollar;b)"
-   *   case i: BigInteger => s"BigInteger(&dollar;i)"
-   *   // drop all other kinds of Number
-   * }
-   * }}}
-   */
-  final case class Widened[T, U](behavior: Behavior[T], matcher: PartialFunction[U, T]) extends ExtensibleBehavior[U] {
-    private def postProcess(behv: Behavior[T], ctx: AC[T]): Behavior[U] =
-      if (isUnhandled(behv)) Unhandled
-      else if (isAlive(behv)) {
-        val next = canonicalize(behv, behavior, ctx)
-        if (next eq behavior) Same else Widened(next, matcher)
-      } else Stopped
-
-    override def receiveSignal(ctx: AC[U], signal: Signal): Behavior[U] =
-      postProcess(Behavior.interpretSignal(behavior, ctx.as[T], signal), ctx.as[T])
-
-    override def receiveMessage(ctx: AC[U], msg: U): Behavior[U] =
-      matcher.applyOrElse(msg, nullFun) match {
-        case null        ⇒ Unhandled
-        case transformed ⇒ postProcess(Behavior.interpretMessage(behavior, ctx.as[T], transformed), ctx.as[T])
-      }
-
-    override def toString: String = s"${behavior.toString}.widen(${LineNumbers(matcher)})"
+    def widen[U](matcher: PartialFunction[U, T]): Behavior[U] =
+      BehaviorImpl.Widened(behavior, matcher)
   }
 
   /**
    * Wrap a behavior factory so that it runs upon PreStart, i.e. behavior creation
    * is deferred to the child actor instead of running within the parent.
    */
-  final case class Deferred[T](factory: ActorContext[T] ⇒ Behavior[T]) extends DeferredBehavior[T] {
-    /** "undefer" the deferred behavior */
-    override def apply(ctx: AC[T]): Behavior[T] = factory(ctx)
-
-    override def toString: String = s"Deferred(${LineNumbers(factory)})"
-  }
+  def Deferred[T](factory: ActorContext[T] ⇒ Behavior[T]): Behavior[T] =
```

Comment:
And always hide the actual implementation behind a method, making it a public API that it is a case class for example seems like a setup for bincomp pain (if there is any such left)